crates/ethrpc/src/buffered.rs

```diff
@@ -84,23 +85,50 @@ where
         tokio::task::spawn(batched_for_each(config, calls, move |batch| {
             let inner = inner.clone();
             async move {
-                let (mut requests, mut senders): (Vec<_>, Vec<_>) = batch
-                    .into_iter()
-                    .filter(|(_, _, sender)| !sender.is_canceled())
-                    .map(|(id, request, sender)| ((id, request), sender))
-                    .unzip();
+                let (mut requests, mut trace_ids, mut senders): (Vec<_>, Vec<_>, Vec<_>) =
+                    itertools::multiunzip(
+                        batch
+                            .into_iter()
+                            .filter(|(_, _, _, sender)| !sender.is_canceled())
+                            .map(|(id, request, trace_id, sender)| {
+                                ((id, request), trace_id, sender)
+                            }),
+                    );
                 match requests.len() {
                     0 => (),
                     1 => {
-                        let ((id, request), sender) = (requests.remove(0), senders.remove(0));
-                        let result = inner.send(id, request).await;
+                        let ((id, request), trace_id, sender) =
+                            (requests.remove(0), trace_ids.remove(0), senders.remove(0));
+                        let result = match trace_id {
+                            Some(trace_id) => {
+                                observe::request_id::set_task_local_storage(
+                                    trace_id,
+                                    inner.send(id, request),
+                                )
+                                .await
+                            }
+                            None => inner.send(id, request).await,
+                        };
                         let _ = sender.send(result);
                     }
                     n => {
-                        let results = inner
-                            .send_batch(requests)
-                            .await
-                            .unwrap_or_else(|err| vec![Err(err); n]);
+                        let request_metadata = requests
+                            .iter()
+                            .zip(trace_ids)
+                            .filter_map(|((_, call), trace_id)| match (call, trace_id) {
+                                (Call::MethodCall(call), Some(trace_id)) => {
+                                    Some(format!("{}:{}", trace_id, call.method))
+                                }
+                                _ => None,
+                            })
+                            .collect_vec()
+                            .join(",");
```

Comment:
I'm slightly worried about the size of these headers. Also currently we can have a lot of duplicate information while at the same time we can lose the ability to associate RPC calls to user requests because we don't have a place holder for requests which were issued outside of a user requests. (currently we can't say with certainty that the nth item in the header list belongs to the nth RPC request in the batch)
I think we should either (1) collect what we have right now into a set to at least reduce duplication or (2) push a placeholder value like `-` for when a request doesn't have a trace id.